### PR TITLE
WIP - DRAFT - Fix merge strategy for NSG/Subnet and remove usage of ID for NSG secl…

### LINKED ISSUE #113: Reconciliation is failing in ClusterClass is Custom NSG/Seclist is used
**What happened**:
Reconciliation is failing in ClusterClass is Custom NSG/Seclist is used. We are hitting the issue in https://github.com/kubernetes-sigs/cluster-api/issues/6320

**What you expected to happen**:
Reconciliation should succeed

**How to reproduce it (as minimally and precisely as possible)**:

Create a ClusterClass template with Custom NSG/Subnet

**Anything else we need to know?**:

**Environment:**

- CAPOCI version: 0.3.0
- Cluster-API version (use `clusterctl version`): 1.1.5
- Kubernetes version (use `kubectl version`):
- Docker version (use `docker info`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
Fix merge strategy for NSG/Subnet and remove usage of ID for NSG seclist reconciliation

**What this PR does / why we need it**:

* NSG/Subnet lists are hitting the issue mentioned https://main.cluster-api.sigs.k8s.io/developer/providers/v1.1-to-v1.2.html#required-api-changes-for-providers . Fix is as per the suggestion in the doc
* NSG Security rules has a similar problem but cant be fixed in a similar way because ID is generated and hence cannot be used in ClusterClass Server side patching. The ClusterClass template will not have the ID. Also, ideally, the NSG Seclist rules should just be  based on the actual rule comparison and not ID comparison. So we have changed code to use rule comparison.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #113

### CODE REVIEW

**Review by joekr — APPROVED** [review]

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -147,14 +147,16 @@ type IngressSecurityRule struct {
 type IngressSecurityRuleForNSG struct {
 	//IngressSecurityRule ID for NSG.
 	// +optional
+	// Deprecated: this field is not populated and used during reconciliation
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply api/v1beta1/types.go:150]: not that I know of, I think we may never be able to remove this. I think that should be ok

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -251,7 +253,6 @@ type Subnet struct {
 	// +optional
 	ID *string `json:"id,omitempty"`
 	// Subnet Name.
-	// +optional
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply api/v1beta1/types.go:254]: the chances are very low, we typically have provided names for subnets in our templates and always sets it in the templates.

[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply api/v1beta1/types.go:254]: also if name is nil, we definitely would have failed reconciliation as we lookup using name and also use it for comparison

**Review by joekr — DISMISSED** [review]

File: templates/cluster-template-cluster-class.yaml [hunk 3]
```diff
@@ -26,3 +26,11 @@ spec:
         value: ${OCI_COMPARTMENT_ID}
       - name: imageId
         value: ${OCI_IMAGE_ID}
+      - name: subnets
+        value:
+        - id: ocid1.subnet.oc1.phx.aaaaaaaa7gyvlsu2vxtpiydb5iu5vg3v3cdb63ayavalabcidbh3sjilm5qq
```
Comment (joekr) [inline templates/cluster-template-cluster-class.yaml:31]: Should we be setting the id here?

File: templates/clusterclass-example.yaml [hunk 4]
```diff
@@ -162,7 +186,17 @@ metadata:
   name: ocicluster
 spec:
   template:
-    spec: {}
+    spec:
+      compartmentId: ocid1.compartment.oc1..aaaaaaaaq6ucotbu72r32xznh5fnh2f77vfk2i35lwazp7gazk6lkzluvwqa
```
Comment (joekr) [inline templates/clusterclass-example.yaml:190]: Should we be using the id here?

**Review by joekr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**